umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,582 @@
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    ensure!(
+        param.signers_and_signatures.len() >= host.state().required_signatures as usize,
+        CustomContractError::NotEnoughSignatures
+    );
+
+    let mut prev_signer: Option<PublicKeyEd25519> = None;
+
+    let message_hash = view_message_hash(ctx, host, crypto_primitives)?;
+
+    let required_signatures = host.state().required_signatures;
+
+    let mut validators: Vec<PublicKeyEd25519> = Vec::with_capacity(required_signatures as usize);
+
+    // To save gas we check only the required number of signatures.
+    // The case, where you can have part of signatures invalid but still enough valid in total is not supported.
+    // We want to record all validators who submit (valid) signatures in a trustless/transparent way
+    // in the smart contract (to e.g. reward off-chain all validators for good behavior) that is why the smart contract allows submitting more signatures than the `required_signatures` here.
+    for i in 0..required_signatures {
+        let signer = param.signers_and_signatures[i as usize].0;
+        let signature = param.signers_and_signatures[i as usize].1;
+
+        //Check signature.
+        let valid_signature =
+            crypto_primitives.verify_ed25519_signature(signer, signature, &message_hash);
+
+        ensure!(valid_signature, CustomContractError::WrongSignature);
+
+        ensure!(
+            prev_signer < Some(signer),
+            CustomContractError::SignaturesOutOfOrder
+        );
+
+        validators.push(signer);
+
+        prev_signer = Some(signer);
+    }
+
+    let are_valid_signers = host.invoke_contract_read_only::<Vec<PublicKeyEd25519>>(
+        &host.state().staking_bank,
+        &validators,
+        EntrypointName::new_unchecked("verifyValidators"),
+        Amount::zero(),
+    )?;
+
+    let are_valid_signers: bool = are_valid_signers
+        .ok_or(CustomContractError::InvokeContractError)?
+        .get()?;
+
+    ensure!(are_valid_signers, CustomContractError::InvalidSigner);
+
+    Ok(())
+}
+
+#[receive(
+    contract = "umbrella_feeds",
+    name = "update",
+    parameter = "UpdateParams",
+    error = "CustomContractError",
+    crypto_primitives,
+    mutable
+)]
+fn update<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    let message = param.message;
+
+    // Check that the signature was intended for this contract.
+    ensure_eq!(
+        message.contract_address,
+        ctx.self_address(),
+        CustomContractError::WrongContract
+    );
+
+    // Check signature is not expired.
+    ensure!(
+        message.timestamp > ctx.metadata().slot_time(),
+        CustomContractError::Expired
```

Comment:
It's a bit odd to call this expired when it's more like "in the future".

Reply:
At the moment this error message is fired. The signature is expired. I feel the naming makes sense so I kept it.